explore/sampling_internal.h

```diff
@@ -0,0 +1,87 @@
+#pragma once
+
+#include <cstdint>
+#include <vector>
+
+namespace exploration
+{
+  const uint64_t a = 0xeece66d5deece66dULL;
+  const uint64_t c = 2147483647;
+
+  const int bias = 127 << 23;
+
+  union int_float
+  {
+    int32_t i;
+    float f;
+  };
+
+  // uniform random between 0 and 1
+  inline float uniform_random_merand48(uint64_t initial)
+  {
+    initial = a * initial + c;
+    int_float temp;
+    temp.i = ((initial >> 25) & 0x7FFFFF) | bias;
+    return temp.f - 1;
+  }
+
+  template<typename InputIt>
+  uint32_t sample_from_pdf(uint64_t seed, InputIt pdf_first, InputIt pdf_last, std::input_iterator_tag pdf_category)
+  {
+    // Create a discrete_distribution based on the returned weights. This class handles the
+    // case where the sum of the weights is < or > 1, by normalizing agains the sum.
+    float total = 0.f;
+    for (InputIt pdf = pdf_first; pdf != pdf_last; ++pdf)
+    {
+      if (*pdf < 0)
+        *pdf = 0;
```

Comment:
Is it valid to do an assignment on an input iterator?

Reply:
It seems odd and unintuitive to have sampling modify the distribution input.  Naively, I would have even imagined that an input iterator disallows that.